File: lib/agents/regulator.ts
// ⚙️ Regulator — Current state agent. Monitors + updates CurrentState every turn.
// The nervous system that connects agents. Always runs first in the Helmsman loop.
import type { HelmContext, CurrentState } from "./state";
import {
  inferMoodFromText,
  inferEnergyFromText,
  moodToTone,
  patchState,
} from "./state";
import type { AgentResult } from "@/lib/types";

function inferPhase(ctx: HelmContext): CurrentState["phase"] {
  const t = ctx.input.toLowerCase();
  if (/break|pause|rest|lunch|walk/.test(t)) return "break";
  if (/plan|what should|priorit|what next|review/.test(t)) return "planning";
  if (/working|focusing|doing|started|in flow/.test(t)) return "working";
  return ctx.state.phase ?? "idle";
}

export function runRegulator(ctx: HelmContext): { state: CurrentState; changed: boolean } {
  const prevState = ctx.state;
  const mood = inferMoodFromText(ctx.input);
  const energy = inferEnergyFromText(ctx.input);
  const toneMode = moodToTone(mood);
  const phase = inferPhase(ctx);

  const newState = patchState(prevState, { mood, energy, toneMode, phase });

  const changed =
    mood !== prevState.mood ||
    Math.abs(energy - prevState.energy) > 0.2 ||
    phase !== prevState.phase;

  return { state: newState, changed };
}

// Registry-compatible wrapper (used by Helmsman before routing)
export async function regulator(ctx: HelmContext): Promise<AgentResult & { newState: CurrentState }> {
  const { state, changed } = runRegulator(ctx);

  const prev = ctx.state;
  const text = changed
    ? `State: ${prev.mood} → ${state.mood} | energy ${(prev.energy * 100).toFixed(0)}% → ${(state.energy * 100).toFixed(0)}% | tone → ${state.toneMode}`
    : `State stable: ${state.mood} | energy ${(state.energy * 100).toFixed(0)}% | tone ${state.toneMode}`;

  return {
    text,
    agent: "regulator",
    newState: state,
    sideEffects: [{
      ts: Date.now(),
      type: "state_update",
      source: "passive" as const,
      value: state,
    }],
  };
}

File: backend/avnik/regulator.py
"""⚙️ Regulator — infers mood/energy/phase from the latest turn. Runs first, always.

Faithful port of lib/agents/regulator.ts.
"""
import re
from .models import CurrentState
from .state import infer_mood_from_text, infer_energy_from_text, mood_to_tone


def _infer_phase(user_input: str, current_phase: str) -> str:
    t = user_input.lower()
    if re.search(r"break|pause|rest|lunch|walk", t):
        return "break"
    if re.search(r"plan|what should|priorit|what next|review", t):
        return "planning"
    if re.search(r"working|focusing|doing|started|in flow", t):
        return "working"
    return current_phase or "idle"


def run_regulator(state: CurrentState, user_input: str) -> CurrentState:
    state.mood = infer_mood_from_text(user_input)
    state.energy = infer_energy_from_text(user_input)
    state.toneMode = mood_to_tone(state.mood)
    state.phase = _infer_phase(user_input, state.phase)
    return state
